federation-2/router-bridge/js-src/plan.ts

```diff
@@ -21,63 +47,463 @@ export class BridgeQueryPlanner {
+// `apollo-tooling` repository.
+//
+// Note that this is not exactly the same algorithm as the operation registry
+// signature function, which continues to live in `apollo-graphql`.
+import {
+  DirectiveNode,
+  FieldNode,
+  FloatValueNode,
+  FragmentDefinitionNode,
+  FragmentSpreadNode,
+  InlineFragmentNode,
+  IntValueNode,
+  ListValueNode,
+  ObjectValueNode,
+  OperationDefinitionNode,
+  print,
+  SelectionSetNode,
+  separateOperations,
+  StringValueNode,
+  visit,
+} from "graphql";
+import sortBy from "lodash.sortby";
+import { Buffer } from "buffer";
+
+export function defaultUsageReportingSignature(
+  ast: DocumentNode,
+  operationName: string
+): string {
+  return printWithReducedWhitespace(
+    sortAST(
+      removeAliases(hideLiterals(dropUnusedDefinitions(ast, operationName)))
+    )
+  );
+}
+
+// Like the graphql-js print function, but deleting whitespace wherever
+// feasible. Specifically, all whitespace (outside of string literals) is
+// reduced to at most one space, and even that space is removed anywhere except
+// for between two alphanumerics.
+//
+// Note that recent versions of graphql-js contain a stripIgnoredCharacters
+// function; it would be better to use that instead, though whenever we change
+// the signature algorithm it does make every operation appear to change in
+// Studio.
+//
+// In a GraphQL AST (which notably does not contain comments), the only place
+// where meaningful whitespace (or double quotes) can exist is in StringNodes.
+// So to print with reduced whitespace, we:
+// - temporarily sanitize strings by replacing their contents with hex
+// - use the default GraphQL printer
+// - minimize the whitespace with a simple regexp replacement
+// - convert strings back to their actual value We normalize all strings to
+//   non-block strings for simplicity.
+// (Unlike stripIgnoredCharacters, this does not remove commas.)
+function printWithReducedWhitespace(ast: DocumentNode): string {
+  const sanitizedAST = visit(ast, {
+    StringValue(node: StringValueNode): StringValueNode {
+      return {
+        ...node,
+        value: Buffer.from(node.value, "utf8").toString("hex"),
+        block: false,
+      };
+    },
+  });
+  const withWhitespace = print(sanitizedAST);
+  const minimizedButStillHex = withWhitespace
+    .replace(/\s+/g, " ")
+    .replace(/([^_a-zA-Z0-9]) /g, (_, c) => c)
+    .replace(/ ([^_a-zA-Z0-9])/g, (_, c) => c);
+  return minimizedButStillHex.replace(/"([a-f0-9]+)"/g, (_, hex) =>
+    JSON.stringify(Buffer.from(hex, "hex").toString("utf8"))
+  );
+}
+
+// sortAST sorts most multi-child nodes alphabetically. Using this as part of
+// your signature calculation function may make it easier to tell the difference
+// between queries that are similar to each other, and if for some reason your
+// GraphQL client generates query strings with elements in nondeterministic
+// order, it can make sure the queries are treated as identical.
```

Comment:
field order is supposed to be significant, should we really have queries with different order reported in the same way? If this is only used for field level reporting that should be fine, but otherwise the queries should be separated, no?

Reply:
I think you're correct, however the code has been copy pasted from the apollo-server implementation in order to make sure the bridge reports usage in the same way as the server

Reply:
alright then